src/transformStateWithClones.js

```diff
@@ -7,7 +7,31 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const res = [];
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        currentState = Object.assign({}, currentState, action.extraData);
+        break;
+
+      case 'removeProperties':
+        currentState = Object.assign({}, currentState);
+
+        for (const key of action.keysToRemove) {
+          delete currentState[key];
+        }
+        break;
+
+      case 'clear':
+        currentState = {};
+        break;
+    }
+    res.push(Object.assign({}, currentState));
```

Comment:
You are creating a new object with Object.assign unnecessarily. Since currentState is already a copy of the state, you can push it directly to the res array without creating another copy.